network/network/src/main/java/bisq/network/p2p/common/RequestHandler.java

```diff
@@ -0,0 +1,94 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public
+ * License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with Bisq. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package bisq.network.p2p.common;
+
+import bisq.network.identity.NetworkId;
+import bisq.network.p2p.message.EnvelopePayloadMessage;
+import bisq.network.p2p.message.Request;
+import bisq.network.p2p.message.Response;
+import bisq.network.p2p.node.CloseReason;
+import bisq.network.p2p.node.Connection;
+import bisq.network.p2p.node.Node;
+
+
+public abstract class RequestHandler<T extends Request, R extends Response> extends BaseHandler implements Node.Listener, LifecycleHandler {
+    protected final Node node;
+    private final RequestHandlerDelegate<T, R> requestHandlerDelegate;
+
+    public RequestHandler(Node node) {
+        this.node = node;
+        this.requestHandlerDelegate = new RequestHandlerDelegate<>(node, getRequestClass(), new RequestHandlerDelegate.Callback<>() {
+            @Override
+            public R createResponse(Connection connection, T request) {
+                return RequestHandler.this.createResponse(connection, request);
+            }
+
+            @Override
+            public void onRequest(Connection connection, T request) {
+                RequestHandler.this.onRequest(connection, request);
+            }
+        });
+    }
+
+    /* --------------------------------------------------------------------- */
+    // RequestHandler implementation
+    /* --------------------------------------------------------------------- */
+
+    protected abstract R createResponse(Connection connection, T request);
+
+    protected abstract Class<T> getRequestClass();
+
+    protected void onRequest(Connection connection, T request) {
+    }
+
+    protected void processRequest(Connection connection, T request) {
+        requestHandlerDelegate.processRequest(connection, request);
+    }
+
+    /* --------------------------------------------------------------------- */
+    // LifecycleHandler implementation
+    /* --------------------------------------------------------------------- */
+
+    public void initialize() {
+        node.addListener(this);
+        requestHandlerDelegate.initialize();
+    }
+
+    public void shutdown() {
+        node.removeListener(this);
+        requestHandlerDelegate.shutdown();
+    }
+
+    /* --------------------------------------------------------------------- */
+    // Node.Listener implementation
+    /* --------------------------------------------------------------------- */
+
+    @Override
+    public void onMessage(EnvelopePayloadMessage envelopePayloadMessage, Connection connection, NetworkId networkId) {
+        requestHandlerDelegate.resolveRequest(envelopePayloadMessage).ifPresent(request -> processRequest(connection, request));
```

Comment:
Small style NIT: We usually place the continuation on streams/optionals in the next line of its not very short.
```
requestHandlerDelegate.resolveRequest(envelopePayloadMessage)
    .ifPresent(request -> processRequest(connection, request));
```